django/conf/service_urls/base.py

```diff
@@ -0,0 +1,89 @@
+import re
+from urllib import parse
+
+
+class Service:
+    validation = re.compile(r'^(?P<scheme>\S+)://\S*')
+
+    def config_from_url(self, engine, scheme, url):
+        raise NotImplementedError('')
+
+    def __init__(self):
+        self._schemes = {}
+
+    def validate(self, data):
+        match = self.validation.match(data)
+        return match.groups()[0] if match else None
+
+    def _parse(self, data):
+        if isinstance(data, str):
+            scheme = self.validate(data)
+            if scheme is None:
+                raise ValueError('{dsn} is invalid, only full dsn urls (scheme://host...) allowed'.format(dsn=data))
+            try:
+                _scheme = self._schemes[scheme]
+            except KeyError:
+                raise ValueError('{scheme}:// scheme not registered'.format(scheme=scheme))
+            callback, engine = _scheme['callback'], _scheme['engine']
+            return callback(self, engine, scheme, data)
+        return data
+
+    def parse(self, data):
+        if isinstance(data, dict):
+            return {k: self._parse(v) for k, v in data.items()}
```

Comment:
Are you sure you don't want to add here:

```
elif isinstance(data, str):
    return self._parse(data)
```

Do you think this would allow to remove the big if block in _parse  ?